scripts/bulk_update_leads_info.py

```diff
@@ -0,0 +1,153 @@
+# -*- coding: utf-8 -*-
+
+import sys
+import re
+import argparse
+import csv
+import logging
+from closeio_api import Client as CloseIO_API, APIError
+
+
+def get_contact_info(contact_no, csv_row, what, contact_type):
+    columns = [x for x in csv_row.keys()
+               if re.match(r'contact%s_%s[0-9]' % (contact_no, what), x) and csv_row[x]]
+    contact_info = []
+    for col in columns:
+        contact_info.append({what: csv_row[col], 'type': contact_type})
+    return contact_info
+
+parser = argparse.ArgumentParser(description='')
+parser.add_argument('csvfile', type=argparse.FileType('rU'), help='csv file')
+parser.add_argument('--api_key', '-k', required=True, help='API Key')
+parser.add_argument('--development', '-d', action='store_true',
+                    help='Use a development (testing) server rather than production.')
+parser.add_argument('--confirmed', '-c', action='store_true',
+                    help='Without this flag, the script will do a dry run without actually updating any data.')
+parser.add_argument('--create-custom-fields', '-C', action='store_true',
+                    help='Create new custom fields, if not exists.')
+parser.add_argument('--disable-create', '-e', action='store_true',
+                    help='Prevent new lead creation. Update only exists leads.')
+args = parser.parse_args()
+
+log_format = "[%(asctime)s] %(levelname)s %(message)s"
+if not args.confirmed:
+    log_format = 'DRY RUN: '+log_format
+logging.basicConfig(level=logging.DEBUG, format=log_format)
+logging.debug('parameters: %s' % vars(args))
+
+sniffer = csv.Sniffer()
+dialect = sniffer.sniff(args.csvfile.read(1024))
+args.csvfile.seek(0)
+c = csv.DictReader(args.csvfile, dialect=dialect)
+assert any(x in ('company', 'lead_id') for x in c.fieldnames), \
+    'ERROR: column "company" or "lead_id" is not found'
+
+
+api = CloseIO_API(args.api_key, development=args.development)
+
+resp = api.get('custom_fields/lead')
+available_custom_fieldnames = [x['name'] for x in resp['data']]
+new_custom_fieldnames = [x for x in [y.split('.')[1] for y in c.fieldnames if y.startswith('custom.')]
+                         if x not in available_custom_fieldnames]
+
+if args.create_custom_fields:
+    for field in new_custom_fieldnames:
+        if args.confirmed:
```

Comment:
I'll rewrite it with abort. Now, without that option I updating only known custom fields. If you will scroll down:

custom = {x.split('.')[1]: r[x] for x in r.keys() if x.startswith('custom.')
              and x.split('.')[1] in available_custom_fieldnames and r[x]}